### ISSUE TITLE
Template based Stacks

### ISSUE THREAD

**Issue by displague** [crossplane/crossplane#853]
## What problem are you facing?

The current design of Stacks allows for two broad permission scopes for Stacks and one implementation approach.

Current Stack Scopes:

* Cluster Scopes
* Namespaced Scopes

Current Implementation Approaches:

* `kind: Deployment` (defined by `install.yaml`)

   The `Deployment` approach requires that some controller exists.  _It is worth noting that the controller a Stack installs can be any controller.  There is no requirement to take advantage of Kubebuilder or the [Crossplane Runtime](https://github.com/crossplaneio/crossplane-runtime)._

   *Crossplane had support for a secondary implementation approach for Stacks that accepted `kind: Job` via `install.yaml`. This approach was lost in late changes for the v0.3 milestone.*

Previous Stack documents have suggested that Stacks would offer an implementation approach that did not require a new controller to be written. One of the suggested approaches was a "**Template based Stack**".  This has not been more concisely defined outside of suggestions.

### Template Based Stacks 

The simplicity and advantages of such a Template based Stack are clear:

* Kubernetes resources can be dynamically revised in response to a parent resource
* Kubernetes resources can be dynamically revised in response to child or sibling resources
* No requirement to learn Go or other programming languages (YAML and a template language not withstanding)

### Issues that need Resolutions

New Stacks, such as the [Wordpress Sample Stack](https://github.com/crossplane/sample-stack-wordpress), should not require a new controller to be written.

* [x] Come up with one or more designs to fulfil this (Controller vs Controllerless, for example) [crossplaneio/crossplane#877]
* [x] Implement the design [crossplaneio/crossplane#878]
* [ ] Document the new feature
* [ ] `crossplane-cli` should be enhanced to work with these Stacks (create, develop, provision, and modify)
* [ ] `crossplane-cli` should be enhanced to work with these Stacks (development scaffolding)
* [ ] Two or more example Stacks should illustrate the new feature and act as a development aid (Relates to #868)
  * [x] Revise the existing Wordpress sample stack or create an alternate Wordpress stack [#881] (or not) 

## How could this be demonstrated
(1) SRE, DevOps, AppOps user: Get Wordpress up and running as a grounding use case:
 - start with yaml: existing application; using wordpress as example; 
   - DB claim
   - workload definition - KubernetesApplication
   - ?cluster claim?
 - run command to init scaffolding `kubectl crossplane stack init --template`
 - put yaml into specific folder
 - update RBAC in app.yaml
 - use all regular stuff to build and publish and run your stack
 - deploy and use app instance
 - cleanly delete app instance and verify all resources have been deleted

(2) SRE, DevOps, AppOps user: Get GitLab up and running as a grounding use case:
 - roughly same as above

## Related
* [ ] [#914] Document key problems that template stacks solve for k8s app deployments - initial draft of key problems we're solving in this epic

* [x] [#915] UX: define the overall experience for template stacks; modeling, design thinking; UX; e.g. #908, plus more; stack example modeling sock shop, whole repo with examples of different options for template stacks experience. get initial experience, then iterate towards best experience.

* [x] #877 Design 1-pager, including systems to make experience possible?
  - registering templates; spawn controllers dynamically at runtime

* [x] #891 Crossplane CLI generating a template stack (should work with #917 when ready)

* [x] #878 Implement template stacks : breakdown into
    - [x] #917 implement template stacks with **no** variables
    - [ ] #918 implement template stacks **with variables**
    - [ ] #919 implement template stacks with **live** variables

* [ ] #881 Build and deploy a working Wordpress as a template stack
  - depends on #917
  - A first template stack experience, for simple templates like Wordpress, that can be re-used
  - create multiple app instances without writing any code?
  - copy the project and re-write the yaml to be your example
  - you would not pass parameters to it, just what Wordpress sample stack does today?

* [ ] #TBD Build and deploy Wordpress as a template stack, with config passed at install time
  - depends on #918
  - second instance with variables being passed

* [ ] #TBD Deploy a new application to showcase using live value injection
  - depends on #919
  - k8s hard way, deployment example, pulling values from external LB, plugging into input of another
 -  variables are dynamically resolved
  - A template stack experience for sophisticated templates
  - arbitrary templates - including value injection

* [ ] #TBD A template stack experience for overlay-style templates (still under consideration)

* [ ] #TBD Getting GitLab template Stack created and published

* [ ] #916 crossplane.io docs for template stacks with working examples

**Comment by suskin** [comment]
I have started a new section called "deliverables", which aims to capture the user-facing experiences that we could create, even if we start with a rudimentary experience and improve it over time. So far, I've only added a few items to it to try to show what I am imagining going into the section. Part of the goal of the section is to focus us on working toward integrating first for specific user scenarios.
